### PR TITLE
add material transparency flags

### PR DESCRIPTION
**Pull request by bryab**
I tried to match your style of dealing with flags.

The problem, however is that I'd like to still be able to check for potentially unknown flags.  For example there is a "stump bark" flag that has unknown purpose, and I'd like to be able to see that a material has that flag.  Any advice here?

### CODE REVIEW

File: src/parser/fragments/material.rs [hunk 1]
```diff
@@ -1,11 +1,65 @@
...
+    pub fn has_transparency(&self) -> bool {
+        self.0 & Self::TRANSPARENCY == Self::TRANSPARENCY
+    }
+}
+
+impl fmt::Debug for TransparencyFlags {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "TransparencyFlags [{:b}]", self.0)
+    }
+}
```
Comment (cjab) [inline src/parser/fragments/material.rs:61]: Nice! Good call, this is handy.

File: src/parser/fragments/material.rs [hunk 2]
```diff
@@ -1,11 +1,65 @@
...
+        self.0.to_le_bytes().to_vec()
+    }
+
+    /// If this material is visible.  If not, this mesh is probably used for collisions only.
+    pub fn is_visible(&self) -> bool {
+        self.0 & Self::VISIBLE == Self::VISIBLE
+    }
+
+    /// This material is masked by a key color
+    pub fn has_mask_or_transparency(&self) -> bool {
```
Comment (cjab) [inline src/parser/fragments/material.rs:37]: Just curious what the `_or_transparency` here means. It kind of made me think this was testing both flags at first glance.

File: src/parser/fragments/material.rs [hunk 3]
```diff
@@ -17,20 +71,20 @@ pub struct MaterialFragment {
...
+    pub transparency_flags: TransparencyFlags,
 
     /// This typically contains 0x004E4E4E but has also bee known to contain 0xB2B2B2.
     /// Could this be an RGB reflectivity value?
     pub params2: u32,
 
-    /// _Unknown_ - Usually contains 0.
-    pub params3: (f32, f32),
+    /// Coordinate in the texture with the mask color
+    pub mask_color_coord: (f32, f32),
```
Comment (cjab) [inline src/parser/fragments/material.rs:87]: You found it! Very cool :D.

**Comment by cjab** [comment]
This is great. Yeah about the unknown flags. I think you should still be able to reach in and see the wrapped `u32` on `TransparencyFlag` with something like: `material.transparency_flags().0`. That _should_ still be public. If not I don't think it's a bad idea to add some sort of method to the `impl` block that either returns the wrapped value _or_ tests arbitrary flags for debugging/exploring. And that's probably something we could do for all of the flag structs if you find something that works well.

File: src/parser/fragments/material.rs [hunk 4]
```diff
@@ -113,9 +168,9 @@ mod tests {
 
         assert_eq!(frag.name_reference, StringReference::new(-22));
         assert_eq!(frag.flags, 0x02);
-        assert_eq!(frag.params1, 0x80000001);
+        assert_eq!(frag.transparency_flags.0, 0x80000001);
```
Comment (cjab) [inline src/parser/fragments/material.rs:171]: Here's what I was I was thinking with just reaching into the wrapped value.

File: src/parser/fragments/material.rs (hunk 2 above)
Comment (bryab) [reply src/parser/fragments/material.rs:37]: Yeah I wasn't sure about the wording here.  I may change the wording around completely actually.  This flag is I think just to indicate this material has some kind of transparency whether using a mask or alpha channel... yeah I will rename some of this.

**Comment by cjab** [comment]
Turns out the inner value _wasn't_ public. I went ahead and implemented `From<TransparencyFlags>` for `u32`. I'm not sure it's the _best_ way to do it but it seems to work! I guess it also could just be public but that wouldn't have given me a chance to play around with From/Into would it? 😆 

https://github.com/cjab/eq_wld/commit/d01576f6243d3a84544ba538dd964cfbe5d5cbc9#diff-d1005ec9ddde9f13c7a3fb417463455821d7816a24a0fac74ba7f1f594582237R71-R76

**Pull request merged**